### ISSUE TITLE
ipywidgets "Error displaying widget" on example pages

### ISSUE THREAD

**Issue by sandertyu** [executablebooks/thebe#226]
On the example pages, such as the [Activate/Status button](https://minrk.github.io/thebelab/activate_button.html) example, ipywidgets displays an error;

![activate-button](https://user-images.githubusercontent.com/59950537/87272784-25fca000-c48c-11ea-9531-a8893d359c13.PNG)

However, when this same code is ran on https://minrk.github.io/thebelab/ , the widget works fine;

![minrk](https://user-images.githubusercontent.com/59950537/87272791-28f79080-c48c-11ea-996a-153a01165404.PNG)

This may not be an issue per se, but I am curious why the widget works on one page but not on the other. I am part of a team which has created a [ckeditor plugin](https://github.com/LibreTexts/ckeditor-binder-plugin) for Libretexts.org to enable users to input live interactive code via ThebeLab. In our ckeditor plugin, the same ipywidgets code as above also gives an "Error displaying widgets" message. I would like to know how to remedy this error displaying issue, as it seems that there is a current fix for it.

Additionally, one of our users is trying to implement the widgets via the `interact()` function but the widgets are not displaying at all (user's full code and libretexts implementation available [here](https://geo.libretexts.org/Courses/University_of_California_Davis/UCD_GEL_56_-_Introduction_to_Geophysics/Geophysics_is_everywhere_in_geology.../1%3A_Rheology_of_Rocks/1.6%3A_Inclined_Plane_Flow_Jupyter_Notebook)). The Issues here are confusing and some hint that fixes have been made, but obviously it is not working for us. The console error messages are exactly the same as outlined in issue #72. There is mention of an [output implementation](https://github.com/minrk/thebelab/issues/72#issuecomment-428922731) in that issue, but no follow up. Any help diagnosing and debugging this output widget issue would be greatly appreciated.

Edit: I've tried to run the widgets on both ThebeLab pages today, and they both appear to work. I'm not sure if something has changed behind the scenes and/or ThebeLab is inconsistent with ipywidgets.

Edit again: Not sure if I'm seeing things, but the ipywidgets are again no longer working on the example code pages. Very strange.

**Comment by celine168** [comment]
After poking around a bit, I only see the issue so far happening on widgets that involve sliders. Other widgets seems to work completely fine.

For example, see this: https://github.com/sandertyu/widgetstuffs/blob/ipywidgets-test/Thebe/ipywidgets.html

When running this in the browser, I get an error with: "Error: Could not create a view for model id ". 
I think the error might have to do with creating the slider widget or child widgets...not sure.

**Comment by celine168** [comment]
As an update to this, calling IntSlider seems to work fine in the current version of Thebe on the master branch. 
I believe that his githubio page may have been using a development version of ipywidgets.

![image](https://user-images.githubusercontent.com/22624020/102813668-0e500e80-437e-11eb-8046-671d1fb0c670.png)

However, calling `interact` is very inconsistent. The slider will sometimes appear, but other times it will flash very quickly for less than a second and disappear immediately afterwards. I also do not get the child view/model view errors anymore.

Here is an image of it appearing fully one time: (which I cannot seem to reproduce)
![ipywidgetscropped](https://user-images.githubusercontent.com/22624020/102813805-5707c780-437e-11eb-84cb-ff47de8a3840.png)

**Comment by sandertyu** [comment]
I was playing with some of the [output widget documentation](https://ipywidgets.readthedocs.io/en/latest/examples/Output%20Widget.html) in both Thebe and JupyterLab and I came across this curious difference;

`ipywidgets.widgets.Output()` in JupyterLab
![jupyterlab](https://cdn.discordapp.com/attachments/763198235237154837/829138307996450827/unknown.png)

`ipywidgets.widgets.Output()` in an HTML page using Thebe
![thebe](https://cdn.discordapp.com/attachments/763198235237154837/829138260881571840/unknown.png)

In JupyterLab, all the calls and modications done to the Output() widget `out` are taking place within the same output "space" (it's all beneath one code cell). In Thebe, each modification to `out` is being rendered within the output of each individual cell. I believe that Thebe is probably instantiating a new Output() widget for each new cell, rather than layering on to a single one as intended.

**Comment by sandertyu** [comment]
The above behavior is present in 0.6.0, but is fixed on the master branch so far as I can tell. It was [this](https://github.com/executablebooks/thebe/commit/beb37928a6d77cfb129b0192f9aa952130f64a2e) commit which made the fix I think. However, ipywidgets is still a little broken as mentioned [here](https://github.com/LibreTexts/metalc/issues/136#issuecomment-816384466)

**Comment by sandertyu** [comment]
I've added [console messages](https://github.com/LibreTexts/thebe/tree/widget-debug) to `manager.js` and `output.js` to try and trace where the remaining issues could be. I used the following code to do a simple test of the widget callback functionality;

```
from ipywidgets import interact, interactive, fixed, interact_manual
import ipywidgets as widgets
def f(x):
    return x
interact(f, x=10);
```

I found that once the widget is initialized and after the first change to the widget (once the cell recognizes that the button has been moved one tick), the [clear_output(wait = false)](https://github.com/LibreTexts/thebe/blob/181df477eab5e41757dccf4a6dca0c4b0906eead/src/output.js#L67) function is called but does not get called afterwards. From there, no visible changes are made to the output of the widget (shown as `11` in the image below). Hopefully the following image is helpful to understand;

![console-log](https://cdn.discordapp.com/attachments/763198235237154837/830964363774787604/clearoutput.png)

**Comment by sandertyu** [comment]
I believe that `output.js` needs to include a [setOutputs()](https://github.com/jupyter-widgets/ipywidgets/blob/e0d41f6f02324596a282bc9e4650fd7ba63c0004/jupyterlab_widgets/src/output.ts#L110) method for the OutputModel class and to include this in the `initialize()` and `defaults()` methods. I'm basing this off of ipywidget's own [output.ts](https://github.com/jupyter-widgets/ipywidgets/blob/e0d41f6f02324596a282bc9e4650fd7ba63c0004/jupyterlab_widgets/src/output.ts). I believe the `setOutputs()` method is important because it includes a call to `clear_output` which Thebe is currently not doing.

**Comment by mandeepika** [comment]
To add on to what @sandertyu explained, we investigated more and put more console logs in the output.js and got the results as displayed in the image below:

![image](https://user-images.githubusercontent.com/67208398/115939210-43668d80-a452-11eb-826c-5e6009be2f49.png)

As is shown in the image, the first time when a change happens and it works, `onMsgIdChange()` function is called. This points out that there was a change in the msgId and the msgType becomes ‘clear_output’ which is when the `clear_output()` function is called. There are several steps which happen after that like the msgType changing to display_data, comm_msg, and so on. But none of this happens in any subsequent changes we make to the slider. Only the `callbacks()` function runs. It never goes to `onMsgIdChange()` and the msgId never gets changed.

**Comment by mandeepika** [comment]
![image](https://user-images.githubusercontent.com/67208398/115939314-96d8db80-a452-11eb-8e4f-3433c3f0c8a5.png)

Additionally, the first change to the slider which actually works goes through another file [services-shim.js](https://github.com/executablebooks/thebe/blob/c1cd0d054f60bfc79fc14be650fc65ac61888cfb/src/services-shim.js) which is the comm API and handles the communication between the front-end and the kernel. As can be seen from the image, it first goes through the [`send()`](https://github.com/executablebooks/thebe/blob/c1cd0d054f60bfc79fc14be650fc65ac61888cfb/src/services-shim.js#L36) function which sends a message to the sibling comm in the backend. Then, it goes to the [`_hookupCallbacks()`](https://github.com/executablebooks/thebe/blob/c1cd0d054f60bfc79fc14be650fc65ac61888cfb/src/services-shim.js#L68) function in the same file which hooks the callback object up with the kernel. But again none of this happens for any subsequent changes to the slider. So maybe the change is not even getting registered with the kernel since the data is not getting sent.

**Comment by moorepants** [comment]
This is fixed by #418.
